File: cobble-binding/cobble-cpp/cpp/structured/write_batch.cpp
#include <cobble/structured/db.hpp>
#include <cobble/structured/single_db.hpp>
#include <cobble/structured/write_batch.hpp>

#include <cstddef>
#include <cstdint>
#include <limits>
#include <type_traits>
#include <utility>
#include <vector>

#include "../detail/error.hpp"
#include "detail/convert.hpp"
#include "detail/impl.hpp"

namespace cobble::structured {
namespace {

constexpr std::uint16_t kHasTtl = 1;
constexpr std::uint16_t kHasFamily = 2;
constexpr std::uint16_t kAwaitDurable = 4;

constexpr std::size_t AlignUp(std::size_t value, std::size_t alignment) {
  return (value + alignment - 1) / alignment * alignment;
}

struct BytesDescriptor {
  std::size_t data;
  std::size_t length;
};

struct WriteOperationDescriptor {
  std::uint8_t operation;
  std::uint8_t kind;
  std::uint16_t bucket;
  std::uint16_t column;
  std::uint16_t flags;
  std::uint32_t ttl_seconds;
  std::uint32_t reserved;
  std::size_t key_data;
  std::size_t key_length;
  std::size_t value_data;
  std::size_t value_length;
  std::size_t elements_data;
  std::size_t element_count;
  std::size_t family_data;
  std::size_t family_length;
};

static_assert(std::is_standard_layout_v<BytesDescriptor>);
static_assert(offsetof(BytesDescriptor, data) == 0);
static_assert(offsetof(BytesDescriptor, length) == sizeof(std::size_t));
static_assert(sizeof(BytesDescriptor) == sizeof(std::size_t) * 2);
static_assert(alignof(BytesDescriptor) == alignof(std::size_t));

constexpr std::size_t kPayloadOffset =
    AlignUp(sizeof(std::uint8_t) * 2 + sizeof(std::uint16_t) * 3 +
                sizeof(std::uint32_t) * 2,
            alignof(std::size_t));
constexpr std::size_t kOperationSize =
    AlignUp(kPayloadOffset + sizeof(std::size_t) * 8, alignof(std::size_t));
static_assert(std::is_standard_layout_v<WriteOperationDescriptor>);
static_assert(offsetof(WriteOperationDescriptor, operation) == 0);
static_assert(offsetof(WriteOperationDescriptor, kind) == 1);
static_assert(offsetof(WriteOperationDescriptor, bucket) == 2);
static_assert(offsetof(WriteOperationDescriptor, column) == 4);
static_assert(offsetof(WriteOperationDescriptor, flags) == 6);
static_assert(offsetof(WriteOperationDescriptor, ttl_seconds) == 8);
static_assert(offsetof(WriteOperationDescriptor, reserved) == 12);
static_assert(offsetof(WriteOperationDescriptor, key_data) == kPayloadOffset);
static_assert(offsetof(WriteOperationDescriptor, key_length) ==
              kPayloadOffset + sizeof(std::size_t));
static_assert(offsetof(WriteOperationDescriptor, value_data) ==
              kPayloadOffset + sizeof(std::size_t) * 2);
static_assert(offsetof(WriteOperationDescriptor, value_length) ==
              kPayloadOffset + sizeof(std::size_t) * 3);
static_assert(offsetof(WriteOperationDescriptor, elements_data) ==
              kPayloadOffset + sizeof(std::size_t) * 4);
static_assert(offsetof(WriteOperationDescriptor, element_count) ==
              kPayloadOffset + sizeof(std::size_t) * 5);
static_assert(offsetof(WriteOperationDescriptor, family_data) ==
              kPayloadOffset + sizeof(std::size_t) * 6);
static_assert(offsetof(WriteOperationDescriptor, family_length) ==
              kPayloadOffset + sizeof(std::size_t) * 7);
static_assert(sizeof(WriteOperationDescriptor) == kOperationSize);
static_assert(alignof(WriteOperationDescriptor) == alignof(std::size_t));

std::size_t Address(BytesView bytes) noexcept {
  return bytes.empty() ? 0 : reinterpret_cast<std::size_t>(bytes.data());
}

std::uint64_t Count(std::size_t value) {
  if (value > std::numeric_limits<std::uint64_t>::max()) {
    throw Error(ErrorCode::kInput, "write operation count exceeds uint64_t");
  }
  return static_cast<std::uint64_t>(value);
}

struct DescriptorStorage {
  std::vector<BytesDescriptor> elements;
  std::vector<WriteOperationDescriptor> operations;
};

DescriptorStorage Describe(std::span<const WriteOperation> input) {
  DescriptorStorage result;
  std::size_t element_count = 0;
  for (const auto &operation : input) {
    if (operation.kind == ColumnKind::kList) {
      if (operation.list_elements.size() >
          std::numeric_limits<std::size_t>::max() - element_count) {
        throw Error(ErrorCode::kInput, "LIST element count overflows size_t");
      }
      element_count += operation.list_elements.size();
    }
  }
  result.elements.reserve(element_count);
  result.operations.reserve(input.size());
  for (const auto &operation : input) {
    const auto operation_tag = static_cast<std::uint8_t>(operation.operation);
    const auto kind_tag = static_cast<std::uint8_t>(operation.kind);
    if (operation_tag > static_cast<std::uint8_t>(WriteOperationType::kDelete))
      throw Error(ErrorCode::kInput, "unknown structured write operation");
    if (kind_tag > static_cast<std::uint8_t>(ColumnKind::kList))
      throw Error(ErrorCode::kInput, "unknown structured column kind");
    if (operation.kind == ColumnKind::kBytes &&
        !operation.list_elements.empty())
      throw Error(ErrorCode::kInput,
                  "BYTES operation cannot contain LIST elements");
    if (operation.kind == ColumnKind::kList && !operation.bytes_value.empty())
      throw Error(ErrorCode::kInput,
                  "LIST operation cannot contain a BYTES value");
    if (operation.operation == WriteOperationType::kDelete &&
        (!operation.bytes_value.empty() || !operation.list_elements.empty()))
      throw Error(ErrorCode::kInput, "delete operation cannot contain a value");

    const auto element_start = result.elements.size();
    for (const auto element : operation.list_elements) {
      result.elements.push_back({Address(element), element.size()});
    }
    const auto &options = operation.options;
    std::uint16_t flags = options.await_durable ? kAwaitDurable : 0;
    if (options.ttl_seconds)
      flags |= kHasTtl;
    if (options.column_family) {
      if (options.column_family->empty())
        throw Error(ErrorCode::kInput, "column family must not be empty");
      flags |= kHasFamily;
    }
    const auto family = options.column_family
                            ? BytesView(reinterpret_cast<const Byte *>(
                                            options.column_family->data()),
                                        options.column_family->size())
                            : BytesView{};
    const auto *elements = operation.list_elements.empty()
                               ? nullptr
                               : result.elements.data() + element_start;
    result.operations.push_back(
        {operation_tag, kind_tag, operation.bucket, operation.column, flags,
         options.ttl_seconds.value_or(0), 0, Address(operation.key),
         operation.key.size(), Address(operation.bytes_value),
         operation.bytes_value.size(), reinterpret_cast<std::size_t>(elements),
         operation.list_elements.size(), Address(family), family.size()});
  }
  return result;
}

std::size_t DescriptorAddress(const DescriptorStorage &storage) noexcept {
  return storage.operations.empty()
             ? 0
             : reinterpret_cast<std::size_t>(storage.operations.data());
}

} // namespace

struct WriteBatch::Impl {
  struct Element {
    std::size_t offset;
    std::size_t length;
  };
  struct Operation {
    WriteOperationType operation;
    ColumnKind kind;
    BucketId bucket;
    ColumnIndex column;
    std::size_t key_offset;
    std::size_t key_length;
    std::size_t value_offset;
    std::size_t value_length;
    std::size_t element_start;
    std::size_t element_count;
    cobble::WriteOptions options;
  };

  std::vector<Byte> arena;
  std::vector<Element> elements;
  std::vector<Operation> operations;
  mutable std::vector<BytesView> element_views;
  mutable std::vector<WriteOperation> borrowed_operations;

  std::pair<std::size_t, std::size_t> Copy(BytesView value) {
    if (value.size() > std::numeric_limits<std::size_t>::max() - arena.size())
      throw Error(ErrorCode::kInput,
                  "WriteBatch byte storage overflows size_t");
    const auto offset = arena.size();
    if (value.empty())
      return {offset, 0};
    arena.insert(arena.end(), value.begin(), value.end());
    return {offset, value.size()};
  }

  BytesView View(std::size_t offset, std::size_t length) const noexcept {
    return length == 0 ? BytesView{} : BytesView(arena.data() + offset, length);
  }

  void Append(WriteOperationType operation, ColumnKind kind, BucketId bucket,
              BytesView key, ColumnIndex column, BytesView value,
              std::span<const BytesView> list,
              const cobble::WriteOptions &options) {
    const auto [key_offset, key_length] = Copy(key);
    const auto [value_offset, value_length] = Copy(value);
    const auto element_start = elements.size();
    for (const auto item : list) {
      const auto [offset, length] = Copy(item);
      elements.push_back({offset, length});
    }
    operations.push_back({operation, kind, bucket, column, key_offset,
                          key_length, value_offset, value_length, element_start,
                          list.size(), options});
  }

  std::span<const WriteOperation> Materialize() const {
    element_views.resize(elements.size());
    for (std::size_t index = 0; index < elements.size(); ++index) {
      element_views[index] =
          View(elements[index].offset, elements[index].length);
    }
    borrowed_operations.resize(operations.size());
    for (std::size_t index = 0; index < operations.size(); ++index) {
      const auto &source = operations[index];
      const auto list = source.element_count == 0
                            ? std::span<const BytesView>{}
                            : std::span<const BytesView>(
                                  element_views.data() + source.element_start,
                                  source.element_count);
      borrowed_operations[index] = {
          source.operation,
          source.kind,
          source.bucket,
          source.column,
          View(source.key_offset, source.key_length),
          View(source.value_offset, source.value_length),
          list,
          source.options};
    }
    return borrowed_operations;
  }

  void Clear() noexcept {
    arena.clear();
    elements.clear();
    operations.clear();
    element_views.clear();
    borrowed_operations.clear();
  }
};

WriteOperation WriteOperation::PutBytes(BucketId bucket, BytesView key,
                                        ColumnIndex column, BytesView value,
                                        cobble::WriteOptions options) {
  return {WriteOperationType::kPut,
          ColumnKind::kBytes,
          bucket,
          column,
          key,
          value,
          {},
          std::move(options)};
}
WriteOperation WriteOperation::PutList(BucketId bucket, BytesView key,
                                       ColumnIndex column,
                                       std::span<const BytesView> elements,
                                       cobble::WriteOptions options) {
  return {WriteOperationType::kPut,
          ColumnKind::kList,
          bucket,
          column,
          key,
          {},
          elements,
          std::move(options)};
}
WriteOperation WriteOperation::MergeBytes(BucketId bucket, BytesView key,
                                          ColumnIndex column, BytesView value,
                                          cobble::WriteOptions options) {
  return {WriteOperationType::kMerge,
          ColumnKind::kBytes,
          bucket,
          column,
          key,
          value,
          {},
          std::move(options)};
}
WriteOperation WriteOperation::MergeList(BucketId bucket, BytesView key,
                                         ColumnIndex column,
                                         std::span<const BytesView> elements,
                                         cobble::WriteOptions options) {
  return {WriteOperationType::kMerge,
          ColumnKind::kList,
          bucket,
          column,
          key,
          {},
          elements,
          std::move(options)};
}
WriteOperation WriteOperation::Delete(BucketId bucket, BytesView key,
                                      ColumnIndex column,
                                      cobble::WriteOptions options) {
  return {WriteOperationType::kDelete,
          ColumnKind::kBytes,
          bucket,
          column,
          key,
          {},
          {},
          std::move(options)};
}

WriteBatch::WriteBatch() : impl_(std::make_unique<Impl>()) {}
WriteBatch::WriteBatch(WriteBatch &&) noexcept = default;
WriteBatch &WriteBatch::operator=(WriteBatch &&) noexcept = default;
WriteBatch::~WriteBatch() = default;

void WriteBatch::PutBytes(BucketId bucket, BytesView key, ColumnIndex column,
                          BytesView value,
                          const cobble::WriteOptions &options) {
  if (!impl_)
    throw Error(ErrorCode::kInvalidState, "WriteBatch has been moved from");
  impl_->Append(WriteOperationType::kPut, ColumnKind::kBytes, bucket, key,
                column, value, {}, options);
}
void WriteBatch::PutList(BucketId bucket, BytesView key, ColumnIndex column,
                         std::span<const BytesView> elements,
                         const cobble::WriteOptions &options) {
  if (!impl_)
    throw Error(ErrorCode::kInvalidState, "WriteBatch has been moved from");
  impl_->Append(WriteOperationType::kPut, ColumnKind::kList, bucket, key,
                column, {}, elements, options);
}
void WriteBatch::MergeBytes(BucketId bucket, BytesView key, ColumnIndex column,
                            BytesView value,
                            const cobble::WriteOptions &options) {
  if (!impl_)
    throw Error(ErrorCode::kInvalidState, "WriteBatch has been moved from");
  impl_->Append(WriteOperationType::kMerge, ColumnKind::kBytes, bucket, key,
                column, value, {}, options);
}
void WriteBatch::MergeList(BucketId bucket, BytesView key, ColumnIndex column,
                           std::span<const BytesView> elements,
                           const cobble::WriteOptions &options) {
  if (!impl_)
    throw Error(ErrorCode::kInvalidState, "WriteBatch has been moved from");
  impl_->Append(WriteOperationType::kMerge, ColumnKind::kList, bucket, key,
                column, {}, elements, options);
}
void WriteBatch::Delete(BucketId bucket, BytesView key, ColumnIndex column,
                        const cobble::WriteOptions &options) {
  if (!impl_)
    throw Error(ErrorCode::kInvalidState, "WriteBatch has been moved from");
  impl_->Append(WriteOperationType::kDelete, ColumnKind::kBytes, bucket, key,
                column, {}, {}, options);
}
void WriteBatch::Clear() noexcept {
  if (impl_)
    impl_->Clear();
}
std::size_t WriteBatch::size() const noexcept {
  return impl_ ? impl_->operations.size() : 0;
}
bool WriteBatch::empty() const noexcept { return size() == 0; }

void Db::Write(std::span<const WriteOperation> operations) const {
  if (!impl_)
    throw Error(ErrorCode::kInvalidState, "structured Db has been moved from");
  const auto descriptors = Describe(operations);
  detail::Translate([&] {
    structured_ffi::native_structured_db_write(*impl_->native,
                                               DescriptorAddress(descriptors),
                                               Count(operations.size()));
  });
}
void Db::Write(WriteBatch &batch) const {
  if (!batch.impl_)
    throw Error(ErrorCode::kInvalidState, "WriteBatch has been moved from");
  Write(batch.impl_->Materialize());
  batch.impl_->Clear();
}
void SingleDb::Write(std::span<const WriteOperation> operations) const {
  if (!impl_)
    throw Error(ErrorCode::kInvalidState,
                "structured SingleDb has been moved from");
  const auto descriptors = Describe(operations);
  detail::Translate([&] {
    structured_ffi::native_structured_single_db_write(
        *impl_->native, DescriptorAddress(descriptors),
        Count(operations.size()));
  });
}
void SingleDb::Write(WriteBatch &batch) const {
  if (!batch.impl_)
    throw Error(ErrorCode::kInvalidState, "WriteBatch has been moved from");
  Write(batch.impl_->Materialize());
  batch.impl_->Clear();
}

} // namespace cobble::structured

File: cobble-binding/cobble-cpp/src/structured/write_batch.rs
use std::{mem, slice, str};

use cobble_binding::structured::{StructuredWriteBatch, StructuredWriteOptions};

use super::conversion::{format_error, input_error};
use super::{BridgeResult, NativeStructuredDb, NativeStructuredSingleDb};

const OP_PUT: u8 = 0;
const OP_MERGE: u8 = 1;
const OP_DELETE: u8 = 2;
const KIND_BYTES: u8 = 0;
const KIND_LIST: u8 = 1;
const FLAG_HAS_TTL: u16 = 1;
const FLAG_HAS_FAMILY: u16 = 2;
const FLAG_AWAIT_DURABLE: u16 = 4;
const KNOWN_FLAGS: u16 = FLAG_HAS_TTL | FLAG_HAS_FAMILY | FLAG_AWAIT_DURABLE;

#[repr(C)]
struct BytesDescriptor {
    data: usize,
    length: usize,
}

#[repr(C)]
struct WriteOperationDescriptor {
    operation: u8,
    kind: u8,
    bucket: u16,
    column: u16,
    flags: u16,
    ttl_seconds: u32,
    reserved: u32,
    key_data: usize,
    key_length: usize,
    value_data: usize,
    value_length: usize,
    elements_data: usize,
    element_count: usize,
    family_data: usize,
    family_length: usize,
}

struct BorrowedOperation<'a> {
    operation: u8,
    kind: u8,
    bucket: u16,
    column: u16,
    key: &'a [u8],
    value: &'a [u8],
    elements: Vec<&'a [u8]>,
    options: StructuredWriteOptions,
}

pub(crate) fn native_structured_db_write(
    db: &NativeStructuredDb,
    descriptor_address: usize,
    count: u64,
) -> BridgeResult<()> {
    // SAFETY: the private C++ wrapper retains every operation and nested payload for this call.
    let operations = unsafe { borrowed_operations(descriptor_address, count)? };
    let mut batch = db.db.new_write_batch();
    apply_operations(&mut batch, operations)?;
    db.db.write_batch(batch).map_err(format_error)
}

pub(crate) fn native_structured_single_db_write(
    db: &NativeStructuredSingleDb,
    descriptor_address: usize,
    count: u64,
) -> BridgeResult<()> {
    // SAFETY: the private C++ wrapper retains every operation and nested payload for this call.
    let operations = unsafe { borrowed_operations(descriptor_address, count)? };
    let mut batch = db.db.new_write_batch();
    apply_operations(&mut batch, operations)?;
    db.db.write_batch(batch).map_err(format_error)
}

fn apply_operations(
    batch: &mut StructuredWriteBatch,
    operations: Vec<BorrowedOperation<'_>>,
) -> BridgeResult<()> {
    for operation in operations {
        match (operation.operation, operation.kind) {
            (OP_PUT, KIND_BYTES) => {
                cobble_binding::structured::ffi::write_batch_put_borrowed_bytes_with_options(
                    batch,
                    operation.bucket,
                    operation.key,
                    operation.column,
                    operation.value,
                    &operation.options,
                )
                .map_err(format_error)?;
            }
            (OP_PUT, KIND_LIST) => {
                cobble_binding::structured::ffi::write_batch_put_borrowed_list_with_options(
                    batch,
                    operation.bucket,
                    operation.key,
                    operation.column,
                    &operation.elements,
                    &operation.options,
                )
                .map_err(format_error)?;
            }
            (OP_MERGE, KIND_BYTES) => {
                cobble_binding::structured::ffi::write_batch_merge_borrowed_bytes_with_options(
                    batch,
                    operation.bucket,
                    operation.key,
                    operation.column,
                    operation.value,
                    &operation.options,
                )
                .map_err(format_error)?;
            }
            (OP_MERGE, KIND_LIST) => {
                cobble_binding::structured::ffi::write_batch_merge_borrowed_list_with_options(
                    batch,
                    operation.bucket,
                    operation.key,
                    operation.column,
                    &operation.elements,
                    &operation.options,
                )
                .map_err(format_error)?;
            }
            (OP_DELETE, _) => batch.delete_with_options(
                operation.bucket,
                operation.key,
                operation.column,
                &operation.options,
            ),
            _ => return Err(input_error("unknown structured write operation")),
        }
    }
    Ok(())
}

/// Decodes exact-layout operation descriptors while borrowing all payload bytes.
///
/// # Safety
///
/// The descriptor array, nested element arrays, and every non-empty byte span must remain readable
/// until this function and the synchronous write call return.
unsafe fn borrowed_operations<'a>(
    descriptor_address: usize,
    count: u64,
) -> BridgeResult<Vec<BorrowedOperation<'a>>> {
    let descriptors = unsafe {
        borrowed_descriptor_slice::<WriteOperationDescriptor>(
            descriptor_address,
            count,
            "write operation",
        )?
    };
    descriptors
        .iter()
        .map(|descriptor| {
            if descriptor.reserved != 0 {
                return Err(input_error("write operation reserved field must be zero"));
            }
            if descriptor.flags & !KNOWN_FLAGS != 0 {
                return Err(input_error("write operation contains unknown option flags"));
            }
            if descriptor.flags & FLAG_HAS_TTL == 0 && descriptor.ttl_seconds != 0 {
                return Err(input_error("TTL seconds require the TTL option flag"));
            }
            if descriptor.operation > OP_DELETE {
                return Err(input_error("unknown structured write operation"));
            }
            if descriptor.kind > KIND_LIST {
                return Err(input_error("unknown structured write value kind"));
            }
            let key =
                unsafe { borrowed_bytes(descriptor.key_data, descriptor.key_length, "write key")? };
            let value = unsafe {
                borrowed_bytes(
                    descriptor.value_data,
                    descriptor.value_length,
                    "write value",
                )?
            };
            let element_descriptors = unsafe {
                borrowed_descriptor_slice::<BytesDescriptor>(
                    descriptor.elements_data,
                    u64::try_from(descriptor.element_count)
                        .map_err(|_| input_error("LIST element count exceeds u64"))?,
                    "LIST element",
                )?
            };
            let elements = element_descriptors
                .iter()
                .map(|element| unsafe {
                    borrowed_bytes(element.data, element.length, "LIST element")
                })
                .collect::<BridgeResult<Vec<_>>>()?;
            let family_bytes = unsafe {
                borrowed_bytes(
                    descriptor.family_data,
                    descriptor.family_length,
                    "column family",
                )?
            };

            if descriptor.operation == OP_DELETE && (!value.is_empty() || !elements.is_empty()) {
                return Err(input_error("delete operation must not contain a value"));
            }
            if descriptor.operation == OP_DELETE && descriptor.kind != KIND_BYTES {
                return Err(input_error("delete operation kind must be BYTES"));
            }
            if descriptor.kind == KIND_BYTES && !elements.is_empty() {
                return Err(input_error(
                    "BYTES operation must not contain LIST elements",
                ));
            }
            if descriptor.kind == KIND_LIST && !value.is_empty() {
                return Err(input_error("LIST operation must not contain a BYTES value"));
            }
            if descriptor.flags & FLAG_HAS_FAMILY == 0 && !family_bytes.is_empty() {
                return Err(input_error(
                    "column family payload requires the family flag",
                ));
            }
            if descriptor.flags & FLAG_HAS_FAMILY != 0 && family_bytes.is_empty() {
                return Err(input_error("column family must not be empty"));
            }

            let mut raw = if descriptor.flags & FLAG_HAS_TTL != 0 {
                cobble_binding::WriteOptions::with_ttl(descriptor.ttl_seconds)
            } else {
                cobble_binding::WriteOptions::default()
            }
            .with_await_durable(descriptor.flags & FLAG_AWAIT_DURABLE != 0);
            if descriptor.flags & FLAG_HAS_FAMILY != 0 {
                raw.column_family = Some(
                    str::from_utf8(family_bytes)
                        .map_err(|_| input_error("column family must be UTF-8"))?
                        .to_owned(),
                );
            }
            Ok(BorrowedOperation {
                operation: descriptor.operation,
                kind: descriptor.kind,
                bucket: descriptor.bucket,
                column: descriptor.column,
                key,
                value,
                elements,
                options: raw.into(),
            })
        })
        .collect()
}

unsafe fn borrowed_descriptor_slice<'a, T>(
    address: usize,
    count: u64,
    name: &str,
) -> BridgeResult<&'a [T]> {
    let count =
        usize::try_from(count).map_err(|_| input_error(format!("{name} count exceeds usize")))?;
    if count == 0 {
        return Ok(&[]);
    }
    if address == 0 {
        return Err(input_error(format!("{name} descriptors must not be null")));
    }
    if !address.is_multiple_of(mem::align_of::<T>()) {
        return Err(input_error(format!("{name} descriptors are not aligned")));
    }
    let byte_len = count
        .checked_mul(mem::size_of::<T>())
        .ok_or_else(|| input_error(format!("{name} descriptor size overflows")))?;
    if byte_len > isize::MAX as usize {
        return Err(input_error(format!(
            "{name} descriptor array exceeds isize::MAX"
        )));
    }
    address
        .checked_add(byte_len)
        .ok_or_else(|| input_error(format!("{name} descriptor address overflows")))?;
    // SAFETY: caller contract plus null/alignment/size/address checks above.
    Ok(unsafe { slice::from_raw_parts(address as *const T, count) })
}

unsafe fn borrowed_bytes<'a>(address: usize, length: usize, name: &str) -> BridgeResult<&'a [u8]> {
    if length == 0 {
        return Ok(&[]);
    }
    if address == 0 {
        return Err(input_error(format!("non-empty {name} must not be null")));
    }
    if length > isize::MAX as usize {
        return Err(input_error(format!("{name} length exceeds isize::MAX")));
    }
    address
        .checked_add(length)
        .ok_or_else(|| input_error(format!("{name} address overflows")))?;
    // SAFETY: caller retains this readable span for the synchronous crossing.
    Ok(unsafe { slice::from_raw_parts(address as *const u8, length) })
}

#[cfg(test)]
mod tests {
    use super::*;

    fn valid_descriptor() -> WriteOperationDescriptor {
        WriteOperationDescriptor {
            operation: OP_PUT,
            kind: KIND_BYTES,
            bucket: 0,
            column: 0,
            flags: 0,
            ttl_seconds: 0,
            reserved: 0,
            key_data: 0,
            key_length: 0,
            value_data: 0,
            value_length: 0,
            elements_data: 0,
            element_count: 0,
            family_data: 0,
            family_length: 0,
        }
    }

    #[test]
    fn write_descriptor_rejects_unknown_flags_and_null_nested_array() {
        let mut descriptor = valid_descriptor();
        descriptor.flags = 8;
        // SAFETY: descriptor points to a live local value.
        let error = unsafe { borrowed_operations((&descriptor as *const _) as usize, 1) }
            .err()
            .unwrap();
        assert!(error.contains("unknown option flags"));

        descriptor.flags = 0;
        descriptor.kind = KIND_LIST;
        descriptor.element_count = 1;
        // SAFETY: descriptor points to a live local value; null nested array is rejected.
        let error = unsafe { borrowed_operations((&descriptor as *const _) as usize, 1) }
            .err()
            .unwrap();
        assert!(error.contains("must not be null"));
    }

    #[test]
    fn write_descriptor_rejects_unflagged_ttl_and_list_delete() {
        let mut descriptor = valid_descriptor();
        descriptor.ttl_seconds = 1;
        // SAFETY: descriptor points to a live local value.
        let error = unsafe { borrowed_operations((&descriptor as *const _) as usize, 1) }
            .err()
            .unwrap();
        assert!(error.contains("TTL option flag"));

        descriptor.ttl_seconds = 0;
        descriptor.operation = OP_DELETE;
        descriptor.kind = KIND_LIST;
        // SAFETY: descriptor points to a live local value.
        let error = unsafe { borrowed_operations((&descriptor as *const _) as usize, 1) }
            .err()
            .unwrap();
        assert!(error.contains("kind must be BYTES"));
    }
}
